### PR TITLE
[importer] polishing importer direct upload (removing extra ajax call from .mako file and using already written code)

### PR DESCRIPTION
**Pull request by agl29**

### CODE REVIEW

File: desktop/libs/indexer/src/indexer/templates/importer.mako [hunk 1]
```diff
@@ -3185,7 +3185,7 @@ ${ commonheader(_("Importer"), "indexer", user, request, "60px") | n,unicode }
           var files = $('#inputfile')[0].files[0];
           fd.append('inputfile', files);
           fd.append('fileFormat', ko.mapping.toJSON(viewModel.createWizard.source));
-  
+          ## $.post ??
```
Comment (agl29) [inline desktop/libs/indexer/src/indexer/templates/importer.mako:3188]: Got no luck on how to remove ajax calls.
looking for the $.post method but was not able to send formData.

Comment (romainr) [reply desktop/libs/indexer/src/indexer/templates/importer.mako:3188]: There are already POST guess type and file format calls, can't we piggy back on those?

Comment (agl29) [reply desktop/libs/indexer/src/indexer/templates/importer.mako]: Done, but with help of a new function (upload_local_file in api3.py). @romainr

File: desktop/libs/indexer/src/indexer/templates/importer.mako [hunk 2]
```diff
@@ -3161,49 +3158,28 @@ ${ commonheader(_("Importer"), "indexer", user, request, "60px") | n,unicode }
...
               processData:false,
-              beforeSend: function (){
-                viewModel.createWizard.isGuessingFormat(true);
-              },
-              success:function (resp) {
-                var newFormat = ko.mapping.fromJS(new FileType(resp['type'], resp));
-                viewModel.createWizard.source.format(newFormat);
-                viewModel.createWizard.isGuessingFormat(false);
+              success:function (response) {
+                viewModel.createWizard.source.path(response['file_url']);
```
Comment (agl29) [inline desktop/libs/indexer/src/indexer/templates/importer.mako:3173]: so when we set the path it automatically activates line no [1790](https://github.com/cloudera/hue/pull/1981/files#diff-febde30edf75f1a484457fb976110629dcddf622992cecdb750b3ecde8ff0377R1790) after this guessFormat and guess_field_type got called automatically ap per the previous code for like s3.

**Comment by huebuild** [comment]
PS: explicit commit/PR names helps

File: desktop/libs/indexer/src/indexer/templates/importer.mako [hunk 3]
```diff
@@ -3161,49 +3158,28 @@ ${ commonheader(_("Importer"), "indexer", user, request, "60px") | n,unicode }
         };
         function upload() {
           var fd = new FormData();
-          fd.append('fileFormat', ko.mapping.toJSON(viewModel.createWizard.source));
+          var files = $('#inputfile')[0].files[0];
+          fd.append('inputfile', files);
+          var file_size = files.size;
+          if (file_size < 30*1000) {
```
Comment (agl29) [inline desktop/libs/indexer/src/indexer/templates/importer.mako:3164]: Tested for small files so, doing good
Screenshot 2021-04-07 at 11 28 25 PM
looking for data_size is less than the 2.5 MB(django.conf.global_settings have this value) then why it is showing error 

but when I am uploading the file of size 800KB then it is showing error (in ss).

**Comment by agl29** [comment]
@romainr  so here we are only tracing the removing the extra code (ajax) from importer.mako file and using the already written code.

File: desktop/libs/indexer/src/indexer/api3.py [hunk 4]
```diff
@@ -228,37 +228,27 @@ def guess_field_types(request):
...
-      elif count <= 5:
-        sample.append(row)
-        csv_data.append(row)
-      else:
-        csv_data.append(row)
+
+    if len(csv_data) <= 5:
+      sample = csv_data[1:]
+    else:
+      sample = csv_data[1:5]
```
Comment (romainr) [inline desktop/libs/indexer/src/indexer/api3.py:235]: Even just this line should work?

File: desktop/libs/indexer/src/indexer/api3.py [hunk 5]
```diff
@@ -228,37 +228,27 @@ def guess_field_types(request):
...
+
+    column_row = csv_data[0]
```
Comment (romainr) [inline desktop/libs/indexer/src/indexer/api3.py:237]: This one won't on empty file or only header :)

File: desktop/libs/indexer/src/indexer/api3.py [hunk 6]
```diff
@@ -228,37 +228,27 @@ def guess_field_types(request):
...
-      if count == 0:
-        column_row = row
-      elif count <= 5:
-        sample.append(row)
-        csv_data.append(row)
-      else:
-        csv_data.append(row)
+
+    if len(csv_data) <= 5:
+      sample = csv_data[1:]
```
Comment (romainr) [inline desktop/libs/indexer/src/indexer/api3.py:233]: In a followup, we could support with/without headers files, cf `hasHeader`

File: desktop/libs/indexer/src/indexer/api3.py [hunk 7]
```diff
@@ -549,7 +539,7 @@ def importer_submit(request):
         request,
         source,
         destination,
-        csv_data,
+        csv_data[1:],
```
Comment (romainr) [inline desktop/libs/indexer/src/indexer/api3.py:542]: Same, if headers only file and no data, or only one line of data

File: desktop/libs/indexer/src/indexer/templates/importer.mako [hunk 8]
```diff
@@ -3161,49 +3158,28 @@ ${ commonheader(_("Importer"), "indexer", user, request, "60px") | n,unicode }
...
-                viewModel.createWizard.source.format(newFormat);
-                viewModel.createWizard.isGuessingFormat(false);
+              success:function (response) {
+                viewModel.createWizard.source.path(response['file_url']);
               }
+            
             });
+          }
+          else {
+            alert("As of now File size must be less than 30KB");
```
Comment (romainr) [inline desktop/libs/indexer/src/indexer/templates/importer.mako:3179]: FYI: we don't use alert but `$.jHueNotify.`

We can investigate the size limit in follow-up (should be more)

File: desktop/libs/indexer/src/indexer/api3.py [hunk 9]
```diff
@@ -737,3 +727,18 @@ def save_pipeline(request):
...
+
+  upload_file = request.FILES['inputfile']
+  fs = FileSystemStorage()
+  name = fs.save(upload_file.name, upload_file)
+  reader = csv.reader(decode_utf8(upload_file))
+
+  csv_data.clear()
+
+  for row in reader:
+    csv_data.append(row)
```
Comment (romainr) [inline desktop/libs/indexer/src/indexer/api3.py:742]: nit: FYI: probably there is a way to add all rows from the reader in one insutruction

actually, I see, you use a global `csv_data`, can we avoid this in a follow-up and reuse the `file_url` instead? (and we can also improve this file with a random key too later)

File: desktop/libs/indexer/src/indexer/templates/importer.mako (hunk 3 above)
Comment (romainr) [inline desktop/libs/indexer/src/indexer/templates/importer.mako:3164]: We should target at least 5MB if not 10MB

**Review by romainr — APPROVED** [review]
Nice improvements!

**Pull request merged**